larpandoracontent/LArControlFlow/StitchingCosmicRayMergingTool.cc

```diff
@@ -590,19 +596,101 @@ void StitchingCosmicRayMergingTool::StitchPfos(const MasterAlgorithm *const pAlg
             }
         }
 
-        for (const ParticleFlowObject *const pPfoToDelete : pfoVector)
+        for (const ParticleFlowObject *const pPfoToDelete : reducedPfoVector)
         {
-            if (pPfoToEnlarge == pPfoToDelete)
+            if (pSelectedPfoToEnlarge == pPfoToDelete)
                 continue;
 
-            pAlgorithm->StitchPfos(pPfoToEnlarge, pPfoToDelete, pfoToLArTPCMap);
-	    stitchedPfosToX0Map.insert(PfoToFloatMap::value_type(pPfoToEnlarge, x0));
+            pAlgorithm->StitchPfos(pSelectedPfoToEnlarge, pPfoToDelete, pfoToLArTPCMap);
+            stitchedPfosToX0Map.insert(PfoToFloatMap::value_type(pSelectedPfoToEnlarge, x0));
         }
     }
 }
 
 //------------------------------------------------------------------------------------------------------------------------------------------
 
+const ParticleFlowObject *StitchingCosmicRayMergingTool::ReduceToLongestStitch(const PfoVector &pfoVector, const ParticleFlowObject *const pPfoToEnlarge,
+    const PfoToLArTPCMap &pfoToLArTPCMap, PfoVector &reducedPfoVector) const
+{
+    if (pfoVector.size() > 2)
+    {
+        this->SelectLongestStitch(pfoVector, pfoToLArTPCMap, reducedPfoVector);
+        std::sort(reducedPfoVector.begin(), reducedPfoVector.end(), LArPfoHelper::SortByNHits);
+
+        return this->GetClosestPfo(pPfoToEnlarge, reducedPfoVector);
+    }
+    else
+    {
+        reducedPfoVector = pfoVector;
+
+        return pPfoToEnlarge;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void StitchingCosmicRayMergingTool::SelectLongestStitch(const PfoVector &pfoVector, const PfoToLArTPCMap &pfoToLArTPCMap, PfoVector &reducedPfoVector) const
+{
+    //ATTN this is the initialization, in case the rest of logic fails
+    reducedPfoVector = pfoVector;
+    reducedPfoVector.resize(2);
+
+    int totalHitsStitched(0);
+
+    for (const ParticleFlowObject *const pPfoToShift1 : pfoVector)
+    {
+        for (const ParticleFlowObject *const pPfoToShift2 : pfoVector)
```

Comment:
Do these two for loops double count the stitches?  I don't think this will cause a problem because you wipe the reducedPfoVector every time, but it would be more efficient to just loop over all pairs of stitches?

Reply:
it is by construction double counted, but only saved once. 
hmm I guess it could be replaced by something like: 
for (iterator it = pfoVector.begin() ; it != pfoVector.end(); ++it) 
   pfo1 = *it;
   for (iterator it2 = it;  ++it2... )

Reply:
Yeah I think you're totally right, it's double counted, but won't change things.  For efficiency purposes I'd suggest doing the change you suggested.  Otherwise the logic looks nice and clean to me and I'm happy with it.  Thanks for doing this!

Reply:
Agreed, the issues at stake are cosmetics and efficiency, with thought about clarity too. I'd have a slight preference for the cosmetic appearance of the initial implementation, but it is of course looping over more combinations of Pfos than is necessary (combinatorics are not big here, though - max one Pfo per volume). What should win is probably the implementation that makes it clearest what you're trying to achieve. Your call!